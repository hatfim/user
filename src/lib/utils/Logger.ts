/* eslint-disable no-console */

const Logger = {
	log(message: string, data?: any): void {
		// Regular log
		console.log(`[LOG] ${message}`, data || '');
	},

	error(message: string, error: Error): void {
		// Log errors with more visibility
		console.error(`[ERROR] ${message}:`, error.message);

		// Optionally log the stack trace for detailed debugging
		console.error(error.stack);
	},

	// You can add more utility methods like warn, info, etc.
};

export default Logger;
