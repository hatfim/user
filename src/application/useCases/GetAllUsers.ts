import Logger from '@lib/utils/Logger';

import { UserDataSchema } from '@user/entities/UserSchema';
import type { UserData } from '@user/entities/UserTypes';
import type { UserRepository } from '@user/repositories/UserRepository';

export interface GetAllUsersUseCase {
	execute(): Promise<UserData[]>;
}

export class GetAllUsers implements GetAllUsersUseCase {
	constructor(private userRepository: UserRepository) {}

	async execute(): Promise<UserData[]> {
		try {
			const users = await this.userRepository.getAllUsers();

			// Validate users against the schema
			if (!UserDataSchema.array().safeParse(users).success) {
				throw new Error('Received invalid user data format');
			}

			return users;
		} catch (error) {
			// Handle errors, perhaps logging or some other side effects.
			if (error instanceof Error) {
				Logger.error('Error fetching users', error);
			} else {
				// Handle or log non-Error thrown values
				Logger.log('An unexpected error occurred:', error);
			}
			throw error;
		}
	}
}
