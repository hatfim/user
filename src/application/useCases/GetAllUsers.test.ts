import Logger from '@lib/utils/Logger';

import { UsersMock } from '@user/__mocks__/UserMockData';
import type { UserRepository } from '@user/repositories/UserRepository';

import { GetAllUsers } from './GetAllUsers';

// Mocking Logger methods for assertion
jest.mock('@lib/utils/logger', () => ({
	error: jest.fn(),
	log: jest.fn(),
}));

// Mocking UserRepository for controlling its behavior
const mockUserRepository: jest.Mocked<UserRepository> = {
	patchUser: jest.fn(),
	deleteUser: jest.fn(),
	getAllUsers: jest.fn(),
};

describe('GetAllUsers UseCase', () => {
	afterEach(() => {
		jest.clearAllMocks();
	});

	it('fetches users successfully', async () => {
		mockUserRepository.getAllUsers.mockResolvedValueOnce(UsersMock);

		const getAllUsersUseCase = new GetAllUsers(mockUserRepository);
		const result = await getAllUsersUseCase.execute();

		expect(result).toEqual(UsersMock);
	});

	it('logs and rethrows error on fetch failure', async () => {
		const error = new Error('Fetch failed');
		mockUserRepository.getAllUsers.mockRejectedValueOnce(error);

		const getAllUsersUseCase = new GetAllUsers(mockUserRepository);

		await expect(getAllUsersUseCase.execute()).rejects.toThrow(error);
		expect(Logger.error).toHaveBeenCalledWith('Error fetching users', error);
	});

	it('logs unexpected error and rethrows', async () => {
		const unexpectedError = 'An unexpected error';
		mockUserRepository.getAllUsers.mockRejectedValueOnce(unexpectedError);

		const getAllUsersUseCase = new GetAllUsers(mockUserRepository);

		await expect(
			getAllUsersUseCase.execute()
		).rejects.toThrowErrorMatchingSnapshot();
		expect(Logger.log).toHaveBeenCalledWith(
			'An unexpected error occurred:',
			unexpectedError
		);
	});

	it('throws error when user data format is invalid', async () => {
		// An example of data that doesn't match the expected format
		const invalidUsersMock = [{ invalidKey: 'invalidValue' }];
		mockUserRepository.getAllUsers.mockResolvedValueOnce(
			invalidUsersMock as any
		);

		const getAllUsersUseCase = new GetAllUsers(mockUserRepository);

		await expect(getAllUsersUseCase.execute()).rejects.toThrow(
			'Received invalid user data format'
		);
		expect(Logger.error).toHaveBeenCalledWith(
			'Error fetching users',
			expect.any(Error)
		);
	});
});
