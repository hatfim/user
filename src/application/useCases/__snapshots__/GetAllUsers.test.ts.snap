// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAllUsers UseCase logs unexpected error and rethrows 1`] = `undefined`;
